### PR TITLE
AV-59563: Some ehancments to the mcp client

### PR DESCRIPTION
**Pull request by sudswasavi**
This patch does the following:
 - Defines some native data structures to start parsing the mcp events for CRDs.
 - Comes up with a mechanism to detect object updates/creates/deletes for
   non-incremental MCP client.

Next steps:
 - Define a queue where each of the objects are pushed for processing by
workers

### CODE REVIEW

File: pkg/istio/mcp/mcp_controller.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	if len(notReady) > 0 {
+		//log.Infof("Configuration not synced: first push for %v not received", notReady)
+		return false
+	}
+	return true
+}
+
+// ConfigDescriptor returns all the ConfigDescriptors that this
+// controller is responsible for
+func (c *Controller) ConfigDescriptor() ConfigDescriptor {
```
Comment (adityats) [inline pkg/istio/mcp/mcp_controller.go]: Do we need this method as it simply returns a known object type?

File: pkg/istio/mcp/mcp_controller.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+		namespace, name := extractNameNamespace(obj.Metadata.Name)
+		fmt.Println("Got an update for name %s", name)
+		if err := schema.Validate(name, namespace, obj.Body); err != nil {
+			// Do not return an error, instead discard the resources so that Pilot can process the rest.
+			//log.Warnf("Discarding incoming MCP resource: validation failed (%s/%s): %v", conf.Namespace, conf.Name, err)
+			continue
+		}
+		createTime := time.Now()
+		if obj.Metadata.CreateTime != nil {
+			var err error
```
Comment (adityats) [inline pkg/istio/mcp/mcp_controller.go]: Pl move this definition out of the for loop

File: pkg/istio/mcp/mcp_controller.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+			Group:             descriptor.Group,
+			Version:           descriptor.Version,
+			Name:              name,
+			Namespace:         namespace,
+			ResourceVersion:   obj.Metadata.Version,
+			CreationTimestamp: createTime,
+			Labels:            obj.Metadata.Labels,
+			Annotations:       obj.Metadata.Annotations,
+			// TODO (sudswas): Change this to take it from config
+			Domain: "avi.internal",
```
Comment (adityats) [inline pkg/istio/mcp/mcp_controller.go]: This would potentially come from the DNS profile on the Cloud?

File: pkg/istio/mcp/schema.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	// IstioConfigTypes lists all Istio config types with schemas and validation
+	IstioConfigTypes = ConfigDescriptor{
+		VirtualService,
+		Gateway,
+		ServiceEntry,
+	}
+)
+
+func ValidateVirtualService(name, namespace string, msg proto.Message) (errs error) {
+	fmt.Println("We will no-op for now")
```
Comment (adityats) [inline pkg/istio/mcp/schema.go]: Pl remove pronouns from all logging, even if it is temporary.

File: pkg/istio/mcp/schema.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+		}
+	}
+	return ProtoSchema{}, false
+}
+
+type ConfigDescriptor []ProtoSchema
+
+type ProtoSchema struct {
+	ClusterScoped bool
+
```
Comment (adityats) [inline pkg/istio/mcp/schema.go]: nit: any reason for extra linebreaks in this definition?

File: pkg/istio/objects/config.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+	// queryable and should be preserved when modifying objects.
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	ResourceVersion string `json:"resourceVersion,omitempty"`
+	// CreationTimestamp records the creation time
+	CreationTimestamp time.Time `json:"creationTimestamp,omitempty"`
+}
+
+func (c ConfigMeta) QueueByTypes(spec proto.Message) {
+	if c.Type == "gateway" {
```
Comment (adityats) [inline pkg/istio/objects/config.go]: Move to switch statements?

File: pkg/istio/objects/config.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	} else if c.Type == "virtual-service" {
+		// Hydrate the VS Structs
+		vs := NewIstioVirtualService(c, spec)
+		key := c.Namespace + ":" + c.Name
+		NewVsMap.AddObj(key, vs)
+	}
+}
+
+func InitializeObjs(objType string) {
+	// Initialize the objects
```
Comment (adityats) [inline pkg/istio/objects/config.go]: same here, pl move to switch statements.

File: pkg/istio/objects/config.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+	if objType == "gateway" {
+		NewGwMap = NewGatewayMap()
+	} else if objType == "virtual-service" {
+		NewVsMap = NewVirtualServiceMap()
+	}
+}
+
+func CalculateUpdates(objType string) {
+	// This method compares the newMap with the old Map and finds out the items to delete/update/create.
```
Comment (adityats) [inline pkg/istio/objects/config.go]: pl resort to multi line comment of /* .. */ for long descriptions.

File: pkg/istio/objects/config.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+	// Eventually newMap replaces the old Map. We look at the old map and compare every key with the new map,
+	// if a key is absent in the new map - we assume it's a candidate for delete.
+	// If the key is present and the resourceVersion is different, we assume it's an update.
+	// If the key is present and the resourceVersion is same, we no-op.
+	// If the key is absent in the old map but present in the newMap - we assume that it's an add operation.
+	if objType == "gateway" {
+		for newKey, newValue := range NewGwMap.gwMap {
```
Comment (adityats) [inline pkg/istio/objects/config.go]: This block of code and the one that follows for objType == 'virtualservice' have exactly the same logic with different variables. Can you pl refactor/compress it ?

File: pkg/istio/objects/istio_gateway.go [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+package objects
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/protobuf/proto"
+)
+
+// Everything is in memory for us, so we will keep things typed on resources despite them containing
```
Comment (adityats) [inline pkg/istio/objects/istio_gateway.go]: Can you pl elaborate with an example here? I am not sure I understand.

**Comment by adityats** [comment]
istio_gateway.go and istio_vs.go seem to have a set of common methods, can they be interfaced in a new file called istio_common.go?

File: pkg/istio/mcp/mcp_controller.go (hunk 3 above)
Comment (sudswasavi) [reply pkg/istio/mcp/mcp_controller.go]: Yeah.

File: pkg/istio/objects/config.go (hunk 8 above)
Comment (sudswasavi) [reply pkg/istio/objects/config.go]: done.

File: pkg/istio/objects/config.go (hunk 6 above)
Comment (sudswasavi) [reply pkg/istio/objects/config.go]: Done.

File: pkg/istio/mcp/schema.go (hunk 5 above)
Comment (sudswasavi) [reply pkg/istio/mcp/schema.go]: done.

File: pkg/istio/mcp/schema.go (hunk 4 above)
Comment (sudswasavi) [reply pkg/istio/mcp/schema.go]: done.

File: pkg/istio/mcp/mcp_controller.go (hunk 2 above)
Comment (sudswasavi) [reply pkg/istio/mcp/mcp_controller.go]: Only the createTime variable can be moved out. We get obj's value only within the loop.

File: pkg/istio/mcp/mcp_controller.go (hunk 1 above)
Comment (sudswasavi) [reply pkg/istio/mcp/mcp_controller.go]: I'd like to keep it this way, it gives a nice abstraction to any method in the future that may want to act upon our supported types.

File: pkg/istio/objects/config.go (hunk 7 above)
Comment (sudswasavi) [reply pkg/istio/objects/config.go]: done.

File: pkg/istio/objects/config.go (hunk 9 above)
Comment (sudswasavi) [reply pkg/istio/objects/config.go]: Let me do this enhancement in the next few iterations as I get some more idea around how these structs would evolve. I will keep it in mind.

File: pkg/istio/objects/istio_gateway.go (hunk 10 above)
Comment (sudswasavi) [reply pkg/istio/objects/istio_gateway.go]: done.

**Comment by sudswasavi** [comment]
@adityats On the Istio_common.go file - sure. Let me evaluate this compression in a couple of iterations. Right now there's an interface that binds them together called ConfigData.

File: pkg/istio/mcp/mcp_controller.go (hunk 1 above)
Comment (adityats) [reply pkg/istio/mcp/mcp_controller.go]: ok

File: pkg/istio/mcp/mcp_controller.go (hunk 2 above)
Comment (adityats) [reply pkg/istio/mcp/mcp_controller.go]: "var err error" will throw an error if there are multiple objects with obj.Metadata.CreateTime field or nil

File: pkg/istio/objects/config.go (hunk 6 above)
Comment (adityats) [reply pkg/istio/objects/config.go]: Missed switch statement here?

File: pkg/istio/objects/config.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+	// Initialize the objects
+	switch objType {
+	case "gateway":
+		NewGwMap = NewIstioObjectMap()
+	case "virtual-service":
+		NewVsMap = NewIstioObjectMap()
+	}
+}
+
+func CalculateUpdates(objType string) {
```
Comment (adityats) [inline pkg/istio/objects/config.go]: This whole method can be a method under ConfigData with the appropriate variables. Pl think about for next set of changes.

File: pkg/istio/objects/object_map.go [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+	// Add a gateway object in this Queue for processing.
+	o.map_lock.RLock()
+	defer o.map_lock.RUnlock()
+	o.objMap[key] = val
+}
+
+/* This has the mechanics to perform actions on the Istio Object. We keep a map to
+figure out what are the operations to be performed against an Istio object.*/
+type IstioObjectOperations struct {
+	opsMap   map[string][]string
```
Comment (adityats) [inline pkg/istio/objects/object_map.go]: any particular format for Map and lock variable naming ?

File: pkg/istio/objects/object_map.go [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+	ops_lock sync.RWMutex
+	utils    *CommonUtils
+}
+
+func NewIstioObjectOperations(objName string) *IstioObjectOperations {
+	opsMap := &IstioObjectOperations{}
+	opsMap.opsMap = make(map[string][]string)
```
Comment (adityats) [inline pkg/istio/objects/object_map.go]: this is a bit confusing, is it possible to rename IstioObjectOperations to something like IstioObjectOpsContainer?
opsCont := <def>; so the usage kind of falls into sync like opsCont.map, opsCont.lock and opsCont.utils?

File: pkg/istio/objects/object_map.go [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+	return obj
+}
+
+// Map Key is - namespace:object_name. This stores already discovered Istio Object Map.
+type IstioObjectMap struct {
+	objMap   map[string]*IstioObject
+	map_lock sync.RWMutex
+}
+
+func NewIstioObjectMap() *IstioObjectMap {
```
Comment (adityats) [inline pkg/istio/objects/object_map.go]: can we rename it without New prefix? I am not sure about the convention here.

Comment (sudswasavi) [reply pkg/istio/objects/object_map.go]: The usual convention in golang for constructor functions is pre-fixing the function with the New key word...Hence the usage.

File: pkg/istio/objects/config.go (hunk 11 above)
Comment (sudswasavi) [reply pkg/istio/objects/config.go]: Sure. Will keep condensing this design as we move forward. I have condensed a few things based on your previous comment in this PR as well.

File: pkg/istio/objects/config.go (hunk 6 above)
Comment (sudswasavi) [reply pkg/istio/objects/config.go]: Done.

File: pkg/istio/mcp/mcp_controller.go (hunk 2 above)
Comment (sudswasavi) [reply pkg/istio/mcp/mcp_controller.go]: Not sure I understood this comment fully. The var err variable is initialised on every loop and each loop processes one object at a time.

File: pkg/istio/objects/object_map.go (hunk 13 above)
Comment (sudswasavi) [reply pkg/istio/objects/object_map.go]: Sure. Done.

**Pull request merged**